network/src/protocols/stream/mod.rs

```diff
@@ -0,0 +1,226 @@
+    stream: Option<InboundStream>,
+}
+
+impl InboundStreamBuffer {
+    pub fn new() -> Self {
+        Self { stream: None }
+    }
+
+    pub fn new_stream(&mut self, header: StreamHeader) -> anyhow::Result<()> {
+        let old = self.stream.replace(InboundStream::new(header)?);
+        if let Some(stream) = old {
+            bail!("Discard existing stream {}", stream.request_id)
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn append_fragment(
+        &mut self,
+        fragment: StreamFragment,
+    ) -> anyhow::Result<Option<NetworkMessage>> {
+        let stream = self
+            .stream
+            .as_mut()
+            .ok_or_else(|| anyhow::anyhow!("No stream exist"))?;
+        let stream_end = stream.append_fragment(fragment)?;
+        if stream_end {
+            Ok(Some(self.stream.take().unwrap().message))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+pub struct InboundStream {
+    request_id: u32,
+    num_fragments: u8,
+    current_fragment_id: u8,
+    message: NetworkMessage,
+}
+
+impl InboundStream {
+    fn new(header: StreamHeader) -> anyhow::Result<Self> {
+        ensure!(
+            !matches!(header.message, NetworkMessage::Error(_)),
+            "Error message is not expected for stream"
+        );
+        Ok(Self {
+            request_id: header.request_id,
+            num_fragments: header.num_fragments,
+            current_fragment_id: 0,
+            message: header.message,
+        })
+    }
+
+    fn append_fragment(&mut self, mut fragment: StreamFragment) -> anyhow::Result<bool> {
+        ensure!(
+            self.request_id == fragment.request_id,
+            "Stream fragment from a different request"
+        );
+        ensure!(
+            self.current_fragment_id + 1 == fragment.fragment_id,
+            "Unexpected fragment id, expected {}, got {}",
+            self.current_fragment_id + 1,
+            fragment.fragment_id
+        );
+        self.current_fragment_id += 1;
+        let raw_data = &mut fragment.raw_data;
+        match &mut self.message {
+            NetworkMessage::Error(_) => panic!("StreamHeader with Error should be rejected"),
+            NetworkMessage::RpcRequest(request) => request.raw_request.append(raw_data),
+            NetworkMessage::RpcResponse(response) => response.raw_response.append(raw_data),
+            NetworkMessage::DirectSendMsg(message) => message.raw_msg.append(raw_data),
+        }
+        Ok(self.current_fragment_id == self.num_fragments)
+    }
+}
+
+pub struct OutboundStream {
+    request_id_gen: U32IdGenerator,
+    max_frame_size: usize,
+    max_message_size: usize,
+    stream_tx: Sender<MultiplexMessage>,
+}
+
+impl OutboundStream {
+    pub fn new(
+        max_frame_size: usize,
+        max_message_size: usize,
+        stream_tx: Sender<MultiplexMessage>,
+    ) -> Self {
+        // some buffer for headers
+        let max_frame_size = max_frame_size - 64;
```

Comment:
nit: can we make this a constant so it's more obvious that we're assuming the headers are <= 64 bytes?

Reply:
Can you get the actual header size instead of assuming 64 bytes?

Reply:
the header size is different depends on the message types, today we only use less than 10